src/Analysis/Engine/Impl/Indexing/SymbolIndex.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Analysis.Indexing {
+    internal class SymbolIndex : ISymbolIndex {
+        private readonly ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>> _index = new ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>>();
+
+        public SymbolIndex() { }
+
+        public IEnumerable<HierarchicalSymbol> HierarchicalDocumentSymbols(Uri uri) {
+            return _index.TryGetValue(uri, out var list) ? list : Enumerable.Empty<HierarchicalSymbol>();
+        }
+
+        public IEnumerable<FlatSymbol> WorkspaceSymbols(string query) {
+            foreach (var kvp in _index) {
+                foreach (var found in WorkspaceSymbolsQuery(query, kvp.Key, kvp.Value)) {
+                    yield return found;
+                }
+            }
+        }
+
+        private IEnumerable<FlatSymbol> WorkspaceSymbolsQuery(string query, Uri uri, IEnumerable<HierarchicalSymbol> symbols) {
+            // Some semblance of a BFS.
+            var ll = new LinkedList<(HierarchicalSymbol, string)>(symbols.Select(s => (s, (string)null)));
```

Comment:
Can you use `Queue<T>` here?

Reply:
Yes, though I avoided doing so to prevent a the Queue's backing array from getting too large. If you don't think that's a problem, I can probably switch to using `TraverseBreadthFirst`.

Reply:
Ah, I remember why I didn't use `TraverseBreadthFirst`; The type I want to traverse is `IEnumerable<HierarchicalSymbol>`, which means that `TraverseBreadthFirst` would need to be given `Func<IEnumerable<HierarchicalSymbol>, IEnumerable<IEnumerable<HierarchicalSymbol>>>`, and even then it wouldn't really be the best BFS where I want to traverse multiple nodes' children at once and flatten them into the same queue.

Reply:
Array can be up to 2^31 items. That should be enough for us. If you care about LOH, I'd prefer couple arrays in LOH than tons of `LinkedListNode` instances in G2.

Reply:
I guess I should rephrase; I don't mean that it will get too large in general, just that it'd grow very quickly as the tree is much, much wider than it is deep and incur a lot of large allocations and copies. I'm happy to change it, just wanted to rephrase my thinking.

Reply:
Switched to a `Queue`.